### PR TITLE
Add policy enforcement for shutting down a container

### PR DESCRIPTION
**Pull request by SeanTAllen**
Adds policy enforcement around shutting down a container. In the supplied framework, shutting down a container is always 
allowed. We update metadata based on the change in state that will be used by other framework rules. This enforcement point
is important for custom policies for metadata tracking and also for allowing the creation of rules like "this container isn't 
allowed to shutdown if some other container is running".

This commit rewrites `signalContainerV2` and gives it a new name. `signalContainerV2` was only called from the kill and
shutdown container functions. Despite only having 2 possible signals it could receive, it accepted any signal. There's a
replacement method called `signalContainerShutdownV2` that has "the same functionality" as `signalContainerV2`. 
`signalContainerShutdownV2` doesn't accept arbitrary signals. Instead, it takes a boolean for whether the shutdown should be
graceful (aka `SIGTERM`) or if we should do a non-graceful shutdown of `SIGKILL`.

This commit aims to keep things "as they are" except for changes that are required for proper and non-evadable policy
enforcement. 

Actual policy enforcement is in a new method `ShutdownContainer` on the `Host` type in `uvm.go`. It is our design goal to 
have all enforcement functions called from within methods in `uvm.go`. 

There's an additional new method on the `Host` as well: `SignalContainerProcess`. `SignalContainerProcess` allows for the 
sending of arbitrary signals from the untrusted host computer to the UVM. The code has been extracted from 
`signalProcessV2` in the bridge and moved onto the `Host` type. The move was required in order to add additional logic 
to `SignalContainerProcess` that `signalProcessV2` lacked.

`SignalContainerProcess` will check to see if the process being signaled is the init process of the container. If it is and the signal
is `SIGTERM` or `SIGKILL` then the shutdown container enforcement rule will be used. We create this "special case" as shutting
down a container is the process of sending `SIGTERM` or `SIGKILL` to the container's init process. The information for whether
a process is the init process for a container is available, but not from the module that bridge is part of, thus the move of 
functionality into `Host`.

The creation of `SignalContainerProcess` was going to be required for when we add support for enforcing policy around
sending arbitrary signals to processes. The current `SignalContainerProcess` was written with that forthcoming changing in 
mind, but doesn't include any logic for the additional enforcement as that will be coming in a commit that will arrive "in
the not so distant future".

Shutdown container policy, because it is always allowed by our framework, doesn't require changing the `securitypolicy` policy
generation tool as there's no user provided input to the new policy rule `shutdown_container`.

When not using an open door policy, a "container not started" error will be returned if a shutdown is attempted on a container 
identifier that wasn't used to start a container.

### CODE REVIEW

File: internal/guest/bridge/bridge_v2.go [hunk 1]
```diff
@@ -229,17 +229,17 @@ func (b *Bridge) shutdownContainerV2(r *Request) (RequestResponse, error) {
...
 
-	return b.signalContainerV2(ctx, span, r, unix.SIGTERM)
+	return b.signalContainerShutdownV2(ctx, span, r, true)
 }
 
-// signalContainerV2 is not a handler func. This is because the actual signal is
-// implied based on the message type of either `killContainerV2` or
-// `shutdownContainerV2`.
-func (b *Bridge) signalContainerV2(ctx context.Context, span *trace.Span, r *Request, signal syscall.Signal) (_ RequestResponse, err error) {
-	defer func() { oc.SetSpanStatus(span, err) }()
```
Comment (matajoh) [inline internal/guest/bridge/bridge_v2.go:239]: Not quite certain what this does, but it is not clear why it should be removed. Just for my edification, why isn't this needed?

File: internal/guest/runtime/hcsv2/uvm.go [hunk 2]
```diff
@@ -435,6 +436,50 @@ func (*Host) Shutdown() {
...
+			return h.ShutdownContainer(ctx, containerID, graceful)
+		}
+	}
+
+	p, err := c.GetProcess(processID)
+	if err != nil {
+		return err
+	}
+
+	return p.Kill(ctx, signal)
```
Comment (matajoh) [inline internal/guest/runtime/hcsv2/uvm.go:480]: I thought this was just the signal container method. Why does it kill the process? Or is that not what this method does?

File: pkg/securitypolicy/framework.rego [hunk 3]
```diff
@@ -207,6 +207,16 @@ exec_in_container := {"matches": matches, "allowed": true} {
...
 
+default shutdown_container := {"allowed": false}
+
+shutdown_container := {"started": started, "allowed": true} {
+    container_started
+    started := {
+        "action": "remove",
+        "key": input.containerID,
+    }
+}
```
Comment (matajoh) [inline pkg/securitypolicy/framework.rego:218]: Should this also remove the entry from `matches` for this container?

File: internal/guest/runtime/hcsv2/uvm.go (hunk 2 above)
Comment (SeanTAllen) [reply internal/guest/runtime/hcsv2/uvm.go:480]: it doesnt "kill the process". it has the same name as the standard unix c api for sending a signal to a process.

File: pkg/securitypolicy/framework.rego (hunk 3 above)
Comment (SeanTAllen) [reply pkg/securitypolicy/framework.rego:218]: should it? if the container is no longer running, should it be removed from matches?

File: internal/guest/bridge/bridge_v2.go (hunk 1 above)
Comment (matajoh) [reply internal/guest/bridge/bridge_v2.go:239]: Gotcha.

File: internal/guest/runtime/hcsv2/uvm.go (hunk 2 above)
Comment (matajoh) [reply internal/guest/runtime/hcsv2/uvm.go:480]: ah, that makes total sense.

**Pull request merged**